### PR TITLE
Update array tests to use Hedgehog

### PR DESCRIPTION
**Pull request by jystic**
This updates the array tests to use the shiny new state machine testing now available in Hedgehog.

Figured it was time to bite the bullet and start trying to roll it out.

! @tranma @amosr 
/jury approved @amosr

### CODE REVIEW

File: icicle-compiler/test/Icicle/Test/Runtime/Array.hs [hunk 1]
```diff
@@ -1,96 +1,359 @@
...
+    n <- forAll $ Gen.int (Range.constant 2 62)
+    2 ^ n === Array.calculateCapacity (ArrayLength (2 ^ n))
 
 prop_roundtrip_writes :: Property
 prop_roundtrip_writes =
-  gamble (listOf (bounded :: Jack Int64)) $ \xs0 ->
-  testIO . bracket Mempool.create Mempool.free $ \pool -> do
-    array <- Array.new pool (fromIntegral (length xs0))
+  property $ do
+    xs0 <- forAll . Gen.list (Range.linear 0 100) $ Gen.int64 Range.linearBounded
```
Comment (tranma) [inline icicle-compiler/test/Icicle/Test/Runtime/Array.hs]: so this is a list that grows linearly from 0 to 100, containing int64s of from minBound..maxBound?
💯 💯 💯

Comment (jystic) [inline icicle-compiler/test/Icicle/Test/Runtime/Array.hs]: yep! (and the int64s grow linearly too, from [0,0] to [minBound,maxBound])

File: icicle-compiler/test/Icicle/Test/Runtime/Array.hs [hunk 2]
```diff
@@ -1,96 +1,359 @@
...
+            new `List.isPrefixOf` old
+      ]
+
+------------------------------------------------------------------------
+
+prop_array_state_machine :: Property
+prop_array_state_machine =
+  property $ do
+    actions <- forAll $
+      Gen.actions (Range.linear 1 100) initialState [newArray, readArray, writeArray, growArray]
```
Comment (tranma) [inline icicle-compiler/test/Icicle/Test/Runtime/Array.hs]: omg! :1st_place_medal: :1st_place_medal: :1st_place_medal:

**Comment by amosr** [comment]
💯

File: icicle-compiler/src/Icicle/Runtime/Data/Array.hs [hunk 3]
```diff
@@ -297,9 +297,24 @@ grow pool arrayOld@(Array ptrOld) lengthNew = do
...
+  if lengthOld > lengthNew then do
+    let
+      deadSize =
+        (lengthOld - lengthNew) * elementSize
+
+      deadPtr =
+        unsafeElementPtr arrayOld `plusPtr`
+        fromIntegral (lengthNew * elementSize)
+
+    c_memset deadPtr 0 (fromIntegral deadSize)
```
Comment (jystic) [inline icicle-compiler/src/Icicle/Runtime/Data/Array.hs]: I actually found a bug, it was expected that when you grow an array, all the newly added elements start off as `0`, this wasn't the case if you shrink an array then grow it again:

```hs
Var 2 = NewArray (ArrayLength 1)
Var 4 = WriteArray (ArrayIndex 0) (-1) (Var 2)
Var 5 = GrowArray (ArrayLength 0) (Var 2)
Var 6 = GrowArray (ArrayLength 1) (Var 5)
Var 7 = ReadArray (ArrayIndex 0) (Var 6) -- reads the -1 value!
```

Comment (amosr) [inline icicle-compiler/src/Icicle/Runtime/Data/Array.hs]: that's really cool! does that mean grow should zero out everything between the old count and the new count?

Comment (jystic) [inline icicle-compiler/src/Icicle/Runtime/Data/Array.hs]: Yeah, so I decided the easiest way to do it was to make sure the unused memory is always initialised to zero. So when we shrink from an array length of say `10` to a length of `5` we need to zero out indices `5` to `9`.

Comment (amosr) [inline icicle-compiler/src/Icicle/Runtime/Data/Array.hs]: sounds good. I don't know, do we even have a way to shrink arrays from source? is there `map_delete`?

Comment (jystic) [inline icicle-compiler/src/Icicle/Runtime/Data/Array.hs]: This is really just for use from Haskell, although I suppose we could use it to test the C code itself.

**Pull request merged**